server/channels/store/sqlstore/reaction_store.go

```diff
@@ -205,24 +207,90 @@ func (s *SqlReactionStore) DeleteAllWithEmojiName(emojiName string) error {
 	return nil
 }
 
-// DeleteOrphanedRows removes entries from Reactions when a corresponding post no longer exists.
-func (s *SqlReactionStore) DeleteOrphanedRows(limit int) (deleted int64, err error) {
-	// We need the extra level of nesting to deal with MySQL's locking
-	const query = `
-	DELETE FROM Reactions WHERE PostId IN (
-		SELECT * FROM (
-			SELECT PostId FROM Reactions
-			LEFT JOIN Posts ON Reactions.PostId = Posts.Id
-			WHERE Posts.Id IS NULL
-			LIMIT ?
-		) AS A
-	)`
-	result, err := s.GetMasterX().Exec(query, limit)
+func (s SqlReactionStore) PermanentDeleteByUser(userId string) error {
+	now := model.GetMillis()
+	postIds := []string{}
+
+	err := s.GetReplicaX().Select(&postIds, "SELECT PostId FROM Reactions WHERE UserId = ?", userId)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get Reactions with userId=%s", userId)
+	}
+
+	batchIds := [][]string{}
+	batchSize := 1000
+	resultsSize := len(postIds)
+	for i := 0; i < resultsSize; i += batchSize {
+		end := i + batchSize
+
+		if end > resultsSize {
+			end = resultsSize
+		}
+
+		batchIds = append(batchIds, postIds[i:end])
+	}
+
+	for _, batch := range batchIds {
+		txn, err := s.GetMasterX().Beginx()
+		if err != nil {
+			return err
+		}
+		defer finalizeTransactionX(txn, &err)
+
+		query := s.getQueryBuilder().
+			Delete("Reactions").
+			Where(sq.And{
+				sq.Eq{"PostId": batch},
+				sq.Eq{"UserId": userId},
+			})
+
+		_, err = txn.ExecBuilder(query)
+		if err != nil {
+			return errors.Wrapf(err, "failed to delete reactions with userId=%s", userId)
+		}
+
+		for _, postId := range batch {
+			_, err = txn.Exec(UpdatePostHasReactionsOnDeleteQuery, now, postId, postId)
```

Comment:
Unfortunately `hasReactions` won't be 0 because there could be other people's reactions to those particular posts. I'm beginning to think that maybe we leave this as a known "Gap" for now when permanently deleting a user? Only issue is how do we clean up these reactions in future?